makelib/golang.mk

```diff
@@ -132,22 +132,37 @@ endif
 GO_COMMON_FLAGS = $(GO_BUILDFLAGS) -tags '$(GO_TAGS)'
 GO_STATIC_FLAGS = $(GO_COMMON_FLAGS) $(GO_PKG_STATIC_FLAGS) -installsuffix static  -ldflags '$(GO_LDFLAGS)'
 
+GO_INIT_REQUIRED_TARGETS = go.vendor.lite
+GO_VENDOR_TARGET = go.vendor
+# switch for go modules
+ifeq ($(GO111MODULE),on)
+# no GOPATH is set, since go modules are being used
+export GOPATH= 
+export GO111MODULE=on
+
+GO_STATIC_FLAGS += -mod=vendor
+GO_INIT_REQUIRED_TARGETS = go.mod.verify
+GO_VENDOR_TARGET = go.mod.vendor
+
+endif
+
 # ====================================================================================
 # Go Targets
 
-go.init: go.vendor.lite
+go.init: $(GO_INIT_REQUIRED_TARGETS)
 	@if ! `$(GO) version | grep -q -E '\bgo($(GO_SUPPORTED_VERSIONS))\b'`; then \
 		$(ERR) unsupported go version. Please make install one of the following supported version: '$(GO_SUPPORTED_VERSIONS)' ;\
 		exit 1 ;\
 	fi
-	@if [ "$(realpath ../../../..)" !=  "$(realpath $(GOPATH))" ]; then \
+	@if [ "$(GO111MODULE)" != "on" ] && [ "$(realpath ../../../..)" !=  "$(realpath $(GOPATH))" ]; then \
```

Comment:
I think you should be using `GO_PKG_DIR` which is writable.

Reply:
With Go modules, you don't get to decide where the modules are installed. By default, they always get installed in `${HOME}/go/pkg/mod`, and this folder is marked as readonly.

Reply:
Since we are using `GOPATH` to build the packages in the work directory instead of user home directory, ` $(GOPATH) != ""` won't be the case, even when when `GO111MODULE=on`